# 2.3 Quality control

Next, you will implement a `FASTQC` quality control step for your input reads.
The inputs are the same as the read pairs used in the `QUANTIFICATION` step.

## 2.3.1 Implementation  

Add the following to the workflow:  
```
workflow {
	...
	fastqc_ch = FASTQC(read_pairs_ch)
}
```

Add the following `FASTQC` process.

> Complete the process by adding the `input`, `tag`, and `publishDir`
> directives

```
process FASTQC {
    tag "FASTQC on $sample_id" // leave empty
	publishDir $params.outdir, mode: 'copy'
	input:
	tuple val(sample_id), path(reads) // leave empty

	output:
	path "fastqc_${sample_id}_logs"

	script:
	"""
	mkdir fastqc_${sample_id}_logs
	fastqc -o fastqc_${sample_id}_logs -f fastq -q ${reads}
	"""
}
```

Run the workflow. 

> What is output?  

```
...

Comand error:
 .command.sh: line 3: fastqc: command not found

...
```

### 2.3.2 Using containers  

This execution will fail because `fastqc` is not installed in your environment.
A docker (**to be replaced with Singularity/Apptainer**) container image with
the remaining software required will be used. 

Create a file called `nextflow.config` in the same directory as `main.nf`. Add
the following and re-run the workflow:
```
process.container = 'nextflow/rnaseq-nf'
docker.enabled = true
docker.runOptions = '-u $(id -u):$(id -g)'
```

This time the execution will work because `docker.enabled = true` tells
nextflow to use the docker container `nextflow/rnaseq-nf`.

Check that the output files have been output correctly in `/results`.  
